ethers-contract/src/stream.rs

```diff
@@ -49,9 +53,140 @@ where
     }
 }
 
+impl<'a, T, R, E> EventStream<'a, T, R, E>
+where
+    T: Stream<Item = Log> + Unpin + 'a,
+    R: 'a,
+    E: 'a,
+{
+    /// This function will attempt to pull events from both event streams. Each
+    /// stream will be polled in a round-robin fashion, and whenever a stream is
+    /// ready to yield an event that event is yielded.
+    ///
+    /// After one of the two event streams completes, the remaining one will be
+    /// polled exclusively. The returned stream completes when both input
+    /// streams have completed.
+    ///
+    ///
+    /// Note that this function consumes both streams and returns a wrapped
+    /// version of them.
+    /// The item of the wrapped stream is an `Either`, and the items that the `self` streams yields
+    /// will be stored in the left-hand variant of that `Either` and the other stream's (`st`) items
+    /// will be wrapped into the right-hand variant of that `Either`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # async fn test<M:ethers_providers::Middleware>(contract: ethers_contract::Contract<M>) {
+    /// # use ethers_core::types::*;
+    /// # use futures_util::stream::StreamExt;
+    /// # use futures_util::future::Either;
+    /// # use ethers_contract::{Contract, ContractFactory, EthEvent};
+    ///
+    /// #[derive(Clone, Debug, EthEvent)]
+    /// pub struct Approval {
+    ///     #[ethevent(indexed)]
+    ///     pub token_owner: Address,
+    ///     #[ethevent(indexed)]
+    ///     pub spender: Address,
+    ///     pub tokens: U256,
+    /// }
+    ///
+    /// #[derive(Clone, Debug, EthEvent)]
+    /// pub struct Transfer {
+    ///     #[ethevent(indexed)]
+    ///     pub from: Address,
+    ///     #[ethevent(indexed)]
+    ///     pub to: Address,
+    ///     pub tokens: U256,
+    /// }
+    ///
+    ///
+    /// let ev1 = contract.event::<Approval>().from_block(1337).to_block(2000);
+    /// let ev2 = contract.event::<Transfer>();
+    ///
+    /// let mut events = ev1.stream().await.unwrap().select(ev2.stream().await.unwrap()).ok();
+    ///
+    /// while let Some(either) = events.next().await {
+    ///     match either {
+    ///         Either::Left(approval) => { let Approval{token_owner,spender,tokens} = approval; }
+    ///         Either::Right(transfer) => { let Transfer{from,to,tokens} = transfer; }
+    ///     }
+    /// }
+    ///
+    /// # }
+    /// ```
+    pub fn select<St>(self, st: St) -> SelectEvent<SelectEither<'a, Result<R, E>, St::Item>>
```

Comment:
Wonder if there's a world where we do this with a macro instead of a function (like the futures `select!`), so that we are not limited to only selecting 2 events?

Reply:
driving multiple streams to completion with `select!` is a bit verbose and the core problem is that we need to return a type that represents all the different events.
you can do this with `EventStream::select` 

```rust
let ev3 = contract.event::<OtherEvent>();

let stream = ev3.stream().await().unwrap().select(
   ev1.stream().await.unwrap().select(ev2.stream().await.unwrap())
);
```
which will return an `Either<OtherEvent, Either<Approval, Transfer>>`

But there'd be also the possibility to create your own event type and use the `Contract::event_with_filter` function

```rust
 #[derive(Debug, Clone, PartialEq, Eq)]
    pub enum Events {
        Approval(Approval),
        Transfer(Transfer),
        Other(Other),
    }
  impl EthLogDecode for Events {...}

```

Reply:
Ah, right! OK if that pattern works then should be fine.